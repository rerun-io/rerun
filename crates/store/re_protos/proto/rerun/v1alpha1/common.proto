syntax = "proto3";

package rerun.common.v1alpha1;

// TODO(cmc): aren't most of the definitions in here dead code?

// supported encoder versions for encoding data
// See `RerunData` and `RerunChunkData` for its usage
enum EncoderVersion {
  ENCODER_VERSION_UNSPECIFIED = 0;
  ENCODER_VERSION_V0 = 1;
}

// RerunChunk is arrow IPC encoded RecordBatch that has
// rerun-specific semantic constraints and can be directly
// converted to a Rerun Chunk (`re_chunk::Chunk`)
message RerunChunk {
  // encoder version used to encode the data
  EncoderVersion encoder_version = 1;

  // Data payload is Arrow IPC encoded RecordBatch
  // TODO(zehiko) make this optional (#9285)
  bytes payload = 2;
}

// unique recording identifier. At this point in time it is the same id as the ChunkStore's StoreId
message RecordingId {
  string id = 1;
}

// uniquely identifies a table
message TableId {
  string id = 1;
}

// A recording can have multiple timelines, each is identified by a name, for example `log_tick`, `log_time`, etc.
message Timeline {
  string name = 1;
}

// A time range between start and end time points. Each 64 bit number can represent different time point data
// depending on the timeline it is associated with. Time range is inclusive for both start and end time points.
message TimeRange {
  int64 start = 1;
  int64 end = 2;
}

// arrow IPC serialized schema
message Schema {
  optional bytes arrow_schema = 1;
}

// TODO(cmc): can we kill this?
message Query {
  // The subset of the database that the query will run on: a set of EntityPath(s) and their
  // associated Component(s)
  ViewContents view_contents = 1;

  // Whether the view_contents should ignore semantically empty columns
  // A semantically empty column is a column that either contains no data at all, or where all
  // values are either nulls or empty arrays ([]).
  bool include_semantically_empty_columns = 2;

  // Whether the view_contents should ignore columns corresponding to indicator components
  // Indicator components are marker components, generally automatically inserted by Rerun, that
  // helps keep track of the original context in which a piece of data was logged/sent.
  bool include_indicator_columns = 3;

  // Whether the view_contents should ignore columns corresponding to Clear-related components
  bool include_tombstone_columns = 4;

  // The index used to filter out _rows_ from the view contents.
  // Only rows where at least 1 column contains non-null data at that index will be kept in the
  // final dataset. If left unspecified, the results will only contain static data.
  IndexColumnSelector filtered_index = 5;

  // The range of index values used to filter out _rows_ from the view contents
  // Only rows where at least 1 of the view-contents contains non-null data within that range will be kept in
  // the final dataset.
  // This has no effect if filtered_index isn't set.
  // This has no effect if using_index_values is set.
  IndexRange filtered_index_range = 6;

  // The specific index values used to filter out _rows_ from the view contents.
  // Only rows where at least 1 column contains non-null data at these specific values will be kept
  // in the final dataset.
  // This has no effect if filtered_index isn't set.
  // This has no effect if using_index_values is set.
  IndexValues filtered_index_values = 7;

  // The specific index values used to sample _rows_ from the view contents.
  // The final dataset will contain one row per sampled index value, regardless of whether data
  // existed for that index value in the view contents.
  // The semantics of the query are consistent with all other settings: the results will be
  // sorted on the filtered_index, and only contain unique index values.
  //
  // This has no effect if filtered_index isn't set.
  // If set, this overrides both filtered_index_range and filtered_index_values.
  IndexValues using_index_values = 8;

  // The component column used to filter out _rows_ from the view contents.
  // Only rows where this column contains non-null data be kept in the final dataset.
  ComponentColumnSelector filtered_is_not_null = 9;

  // The specific _columns_ to sample from the final view contents.
  // The order of the samples will be respected in the final result.
  //
  // If unspecified, it means - everything.
  ColumnSelection column_selection = 10;

  // Specifies how null values should be filled in the returned dataframe.
  SparseFillStrategy sparse_fill_strategy = 11;

  reserved 12;
  reserved "include_properties_entity";
}

message ColumnSelection {
  repeated ColumnSelector columns = 1;
}

message ColumnSelector {
  oneof selector_type {
    ComponentColumnSelector component_column = 2;
    TimeColumnSelector time_column = 3;
  }
}

message IndexColumnSelector {
  // TODO(zehiko) we need to add support for other types of index selectors
  Timeline timeline = 1;
}

message IndexRange {
  // TODO(zehiko) support for other ranges for other index selectors
  TimeRange time_range = 1;
}

message IndexValues {
  // TODO(zehiko) we need to add support for other types of index selectors
  repeated TimeInt time_points = 1;
}

message SampledIndexValues {
  repeated TimeInt sample_points = 1;
}

// A 64-bit number describing either nanoseconds, sequence numbers or fully static data.
message TimeInt {
  int64 time = 1;
}

message ViewContents {
  repeated ViewContentsPart contents = 1;
}

message ViewContentsPart {
  EntityPath path = 1;
  ComponentsSet components = 2;
}

message ComponentsSet {
  repeated Component components = 1;
}

// The unique identifier of an entity, e.g. `camera/3/points`
// See <https://www.rerun.io/docs/concepts/entity-path> for more on entity paths.
message EntityPath {
  string path = 1;
}

// Component describes semantic data that can be used by any number of  rerun's archetypes.
message Component {
  // component name needs to be a string as user can define their own component
  string name = 1;
}

// Used to telect a time column.
message TimeColumnSelector {
  Timeline timeline = 1;
}

// Used to select a component based on its EntityPath and Component name.
message ComponentColumnSelector {
  EntityPath entity_path = 1;
  Component component = 2;
}

// Specifies how null values should be filled in the returned dataframe.
enum SparseFillStrategy {
  SPARSE_FILL_STRATEGY_UNSPECIFIED = 0;
  SPARSE_FILL_STRATEGY_NONE = 1;
  SPARSE_FILL_STRATEGY_LATEST_AT_GLOBAL = 2;
}

message ApplicationId {
  string id = 1;
}

enum StoreKind {
  STORE_KIND_UNSPECIFIED = 0;
  STORE_KIND_RECORDING = 1;
  STORE_KIND_BLUEPRINT = 2;
}

message StoreId {
  StoreKind kind = 1;
  string id = 2;
}

message Tuid {
  // Approximate nanoseconds since epoch.
  optional fixed64 time_ns = 1;

  // Initialized to something random on each thread, then incremented for each
  // new `Tuid` being allocated.
  optional fixed64 inc = 2;
}

message EntryId {
  Tuid id = 1;
}

// Entry point for all ManifestRegistryService APIs
message DatasetHandle {
  // Unique entry identifier (for debug purposes)
  EntryId entry_id = 1;

  // Path to Dataset backing storage (e.g. s3://bucket/file or file:///path/to/file)
  optional string dataset_url = 2;
}

// DataframePart is arrow IPC encoded RecordBatch
message DataframePart {
  // encoder version used to encode the data
  EncoderVersion encoder_version = 1;

  // Data payload is Arrow IPC encoded RecordBatch
  optional bytes payload = 2;
}

// Generic parameters that will influence the behavior of the Lance scanner.
//
// TODO(zehiko, cmc): This should be available for every endpoint that queries data in
// one way or another.
message ScanParameters {
  // List of columns to project. If empty, all columns will be projected.
  repeated string columns = 1;

  IfMissingBehavior on_missing_columns = 2;

  // An arbitrary filter expression that will be passed to the Lance scanner as-is.
  //
  // ```text
  // scanner.filter(filter)
  // ```
  optional string filter = 3;

  // An arbitrary offset that will be passed to the Lance scanner as-is.
  //
  // ```text
  // scanner.limit(_, limit_offset)
  // ```
  optional int64 limit_offset = 4;

  // An arbitrary limit that will be passed to the Lance scanner as-is.
  //
  // ```text
  // scanner.limit(limit_len, _)
  // ```
  optional int64 limit_len = 5;

  // An arbitrary order clause that will be passed to the Lance scanner as-is.
  //
  // ```text
  // scanner.order_by(â€¦)
  // ```
  ScanParametersOrderClause order_by = 6;

  // If set, the output of `scanner.explain_plan` will be dumped to the server's log.
  bool explain_plan = 7;

  // If set, the final `scanner.filter` will be dumped to the server's log.
  bool explain_filter = 8;
}

enum IfMissingBehavior {
  IF_MISSING_BEHAVIOR_UNSPECIFIED = 0;
  IF_MISSING_BEHAVIOR_SKIP = 1;
  IF_MISSING_BEHAVIOR_ERROR = 2;
}

// Specify how the relevant creation call behaves
// in case of previously created (duplicate) items
enum IfDuplicateBehavior {
  IF_DUPLICATE_BEHAVIOR_UNSPECIFIED = 0;

  // Overwrite the existing item
  IF_DUPLICATE_BEHAVIOR_OVERWRITE = 1;

  // Skip if the item already exists
  IF_DUPLICATE_BEHAVIOR_SKIP = 2;

  // Return an error if the item already exists
  IF_DUPLICATE_BEHAVIOR_ERROR = 3;
}

message ScanParametersOrderClause {
  bool descending = 1;
  bool nulls_last = 2;
  optional string column_name = 3;
}

// Unique identifier for a partition. Can be user defined
// which means it can be of any type. For simplicity we start
// with a string, but we will probably revisit this.
message PartitionId {
  optional string id = 1;
}

message ComponentDescriptor {
  // Optional name of the `Archetype` associated with this data.
  optional string archetype_name = 1;

  // Optional name of the field within `Archetype` associated with this data.
  optional string archetype_field_name = 2;

  // Semantic name associated with this data.
  optional string component_name = 3;
}

// Unique identifier of a task submitted in the redap
// tasks subsystem
message TaskId {
  string id = 1;
}
