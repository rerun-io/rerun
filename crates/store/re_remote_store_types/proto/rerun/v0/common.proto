syntax = "proto3";

package rerun.remote_store.v0;

// unique recording identifier. At this point in time it is the same id as the ChunkStore's StoreId
message RecordingId {
    string id = 1;
}

// A recording can have multiple timelines, each is identified by a name, for example `log_tick`, `log_time`, etc.
message Timeline {
    string name = 1;
}

// A time range between start and end time points. Each 64 bit number can represent different time point data
// depending on the timeline it is associated with. Time range is inclusive for both start and end time points.
message TimeRange {
    int64 start = 1;
    int64 end = 2;
}

// arrow IPC serialized schema
message Schema {
    bytes arrow_schema = 1;
}

message Query {
    // The subset of the database that the query will run on: a set of EntityPath(s) and their
    // associated Component(s)
    ViewContents view_contents = 1;

    // Whether the view_contents should ignore semantically empty columns
    // A semantically empty column is a column that either contains no data at all, or where all
    // values are either nulls or empty arrays ([]).
    bool include_semantically_empty_columns = 2;

    // Whether the view_contents should ignore columns corresponding to indicator components
    // Indicator components are marker components, generally automatically inserted by Rerun, that
    // helps keep track of the original context in which a piece of data was logged/sent.
    bool include_indicator_columns = 3;

    // Whether the view_contents should ignore columns corresponding to Clear-related components
    bool include_tombstone_columns = 4;

    // The index used to filter out _rows_ from the view contents.
    // Only rows where at least 1 column contains non-null data at that index will be kept in the
    // final dataset. If left unspecified, the results will only contain static data.
    IndexColumnSelector filtered_index = 5;

    // The range of index values used to filter out _rows_ from the view contents
    // Only rows where at least 1 of the view-contents contains non-null data within that range will be kept in
    // the final dataset.
    // This has no effect if filtered_index isn't set.
    // This has no effect if using_index_values is set.
    IndexRange filtered_index_range = 6;

    // The specific index values used to filter out _rows_ from the view contents.
    // Only rows where at least 1 column contains non-null data at these specific values will be kept
    // in the final dataset.
    // This has no effect if filtered_index isn't set.
    // This has no effect if using_index_values is set.
    IndexValues filtered_index_values = 7;

    // The specific index values used to sample _rows_ from the view contents.
    // The final dataset will contain one row per sampled index value, regardless of whether data
    // existed for that index value in the view contents.
    // The semantics of the query are consistent with all other settings: the results will be
    // sorted on the filtered_index, and only contain unique index values.
    //
    // This has no effect if filtered_index isn't set.
    // If set, this overrides both filtered_index_range and filtered_index_values.
    IndexValues using_index_values = 8;

    // The component column used to filter out _rows_ from the view contents.
    // Only rows where this column contains non-null data be kept in the final dataset.
    ComponentColumnSelector filtered_is_not_null = 9;

    /// The specific _columns_ to sample from the final view contents.
    /// The order of the samples will be respected in the final result.
    ///
    /// If unspecified, it means - everything.
    ColumnSelection column_selection = 10;

    // Specifies how null values should be filled in the returned dataframe.
    SparseFillStrategy sparse_fill_strategy = 11;
}

message ColumnSelection {
    repeated ColumnSelector columns = 1;
}

message ColumnSelector {
    oneof selector_type {
        ComponentColumnSelector component_column = 2;
        TimeColumnSelector time_column = 3;
    }
}

message IndexColumnSelector {
    // TODO(zehiko) we need to add support for other types of index selectors
    Timeline timeline = 1;
}

message IndexRange {
    // TODO(zehiko) support for other ranges for other index selectors
    TimeRange time_range = 1;
}

message IndexValues {
    // TODO(zehiko) we need to add support for other types of index selectors
    repeated TimeInt time_points = 1;
}

message SampledIndexValues {
    repeated TimeInt sample_points = 1;
}

// A 64-bit number describing either nanoseconds, sequence numbers or fully static data.
message TimeInt {
    int64 time = 1;
}

message ViewContents {
    repeated ViewContentsPart contents = 1;
}

message ViewContentsPart {
    EntityPath path = 1;
    ComponentsSet components = 2;
}

message ComponentsSet {
    repeated Component components = 1;
}

// The unique identifier of an entity, e.g. `camera/3/points`
// See <https://www.rerun.io/docs/concepts/entity-path> for more on entity paths.
message EntityPath {
    string path = 1;
}

// Component describes semantic data that can be used by any number of  rerun's archetypes.
message Component {
    // component name needs to be a string as user can define their own component
    string name = 1;
}

// Used to telect a time column.
message TimeColumnSelector {
    Timeline timeline = 1;
}

// Used to select a component based on its EntityPath and Component name.
message ComponentColumnSelector {
    EntityPath entity_path = 1;
    Component component = 2;
}

// Specifies how null values should be filled in the returned dataframe.
enum SparseFillStrategy {
    NONE = 0;
    LATEST_AT_GLOBAL = 1;
}
