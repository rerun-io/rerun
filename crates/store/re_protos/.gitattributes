src/v0/** linguist-generated=true
