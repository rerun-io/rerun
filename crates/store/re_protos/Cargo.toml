[package]
name = "re_protos"
authors.workspace = true
description = "Rerun remote gRPC/protobuf API types"
edition.workspace = true
homepage.workspace = true
include.workspace = true
license.workspace = true
publish = true
readme = "README.md"
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[dependencies]
re_log_types.workspace = true
re_dataframe.workspace = true

# External
arrow2 = { workspace = true, features = ["io_ipc"] }
prost.workspace = true
thiserror.workspace = true

# Native dependencies:
[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
tonic = { workspace = true, default-features = false, features = [
  "codegen",
  "prost",
  "transport",
] }

# Web dependencies:
[target.'cfg(target_arch = "wasm32")'.dependencies]
tonic = { workspace = true, default-features = false, features = [
  "codegen",
  "prost",
] }
tonic-web-wasm-client.workspace = true

[lints]
workspace = true
