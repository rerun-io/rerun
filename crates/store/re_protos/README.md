# re_protos

Part of the [`rerun`](https://github.com/rerun-io/rerun) family of crates.

[![Latest version](https://img.shields.io/crates/v/re_protos.svg)](https://crates.io/crates/re_protos?speculative-link)
[![Documentation](https://docs.rs/re_protos/badge.svg)](https://docs.rs/re_protos?speculative-link)
![MIT](https://img.shields.io/badge/license-MIT-blue.svg)
![Apache](https://img.shields.io/badge/license-Apache-blue.svg)

Rerun remote store node gRPC API service types (client and server).

This crate includes both the language-agnostic definitions (protobuf) as well as the generated code.

The code is generated with `pixi run codegen-rstore`.
