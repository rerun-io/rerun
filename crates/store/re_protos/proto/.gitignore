schema_snapshot.main.yaml
