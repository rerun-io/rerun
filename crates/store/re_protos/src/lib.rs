//! This crate contains generated types for the remote store gRPC service API.
//! Generation is done using the `re_protos_builder` crate.
//!
//! We want clear separation between 'internal' types and gRPC types and don't want
//! to use gRPC types in the rerun viewer codebase. That's why we implement all the
//! necessary conversion code (in the form of `From` and `TryFrom` traits) in this crate.
//!

pub mod external {
    pub use prost;
}

// This extra module is needed, because of how imports from different packages are resolved.
// For example, `rerun.remote_store.v0.EncoderVersion` is resolved to `super::super::remote_store::v0::EncoderVersion`.
// We need an extra module in the path to `common` to make that work.
// Additionally, the `common` module itself has to exist with a `v0` module inside of it,
// which is the reason for the `common`, `log_msg`, `remote_store`, etc. modules below.

// Note: Be careful with `#[path]` attributes: https://github.com/rust-lang/rust/issues/35016
mod v0 {
    // Note: `allow(clippy::all)` does NOT allow all lints
    #![allow(clippy::all, clippy::pedantic, clippy::nursery)]

    #[path = "./rerun.common.v0.rs"]
    pub mod rerun_common_v0;

    #[path = "./rerun.log_msg.v0.rs"]
    pub mod rerun_log_msg_v0;

    #[path = "./rerun.remote_store.v0.rs"]
    pub mod rerun_remote_store_v0;

    #[path = "./rerun.sdk_comms.v0.rs"]
    pub mod rerun_sdk_comms_v0;
}

pub mod common {
    pub mod v0 {
        pub use crate::v0::rerun_common_v0::*;
    }
}

pub mod log_msg {
    pub mod v0 {
        pub use crate::v0::rerun_log_msg_v0::*;
    }
}

/// Generated types for the remote store gRPC service API v0.
pub mod remote_store {

    pub mod v0 {
        pub use crate::v0::rerun_remote_store_v0::*;

        /// Recording catalog mandatory field names. All mandatory metadata fields are prefixed
        /// with "rerun_" to avoid conflicts with user-defined fields.
        pub const CATALOG_ID_FIELD_NAME: &str = "rerun_recording_id";
        pub const CATALOG_APP_ID_FIELD_NAME: &str = "rerun_application_id";
        pub const CATALOG_START_TIME_FIELD_NAME: &str = "rerun_start_time";
        pub const CATALOG_DESCRIPTION_FIELD_NAME: &str = "rerun_description";
        pub const CATALOG_RECORDING_TYPE_FIELD_NAME: &str = "rerun_recording_type";
        pub const CATALOG_STORAGE_URL_FIELD_NAME: &str = "rerun_storage_url";
        pub const CATALOG_REGISTRATION_TIME_FIELD_NAME: &str = "rerun_registration_time";
        pub const CATALOG_ROW_ID_FIELD_NAME: &str = "rerun_row_id";
    }
}

pub mod sdk_comms {
    pub mod v0 {
        pub use crate::v0::rerun_sdk_comms_v0::*;
    }
}

#[derive(Debug, thiserror::Error)]
pub enum TypeConversionError {
    #[error("missing required field: {package_name}.{type_name}.{field_name}")]
    MissingField {
        package_name: &'static str,
        type_name: &'static str,
        field_name: &'static str,
    },

    #[error("invalid value for field {package_name}.{type_name}.{field_name}: {reason}")]
    InvalidField {
        package_name: &'static str,
        type_name: &'static str,
        field_name: &'static str,
        reason: String,
    },

    #[error("failed to decode: {0}")]
    DecodeError(#[from] prost::DecodeError),

    #[error("failed to encode: {0}")]
    EncodeError(#[from] prost::EncodeError),

    #[error("{0}")]
    UnknownEnumValue(#[from] prost::UnknownEnumValue),
}

impl TypeConversionError {
    #[inline]
    pub fn missing_field<T: prost::Name>(field_name: &'static str) -> Self {
        Self::MissingField {
            package_name: T::PACKAGE,
            type_name: T::NAME,
            field_name,
        }
    }

    #[allow(clippy::needless_pass_by_value)] // false-positive
    #[inline]
    pub fn invalid_field<T: prost::Name>(field_name: &'static str, reason: &impl ToString) -> Self {
        Self::InvalidField {
            package_name: T::PACKAGE,
            type_name: T::NAME,
            field_name,
            reason: reason.to_string(),
        }
    }
}

#[macro_export]
macro_rules! missing_field {
    ($type:ty, $field:expr $(,)?) => {
        $crate::TypeConversionError::missing_field::<$type>($field)
    };
}

#[macro_export]
macro_rules! invalid_field {
    ($type:ty, $field:expr, $reason:expr $(,)?) => {
        $crate::TypeConversionError::invalid_field::<$type>($field, &$reason)
    };
}

mod sizes {
    use re_byte_size::SizeBytes;

    impl SizeBytes for crate::log_msg::v0::LogMsg {
        #[inline]
        fn heap_size_bytes(&self) -> u64 {
            let Self { msg } = self;

            match msg {
                Some(msg) => msg.heap_size_bytes(),
                None => 0,
            }
        }
    }

    impl SizeBytes for crate::log_msg::v0::log_msg::Msg {
        #[inline]
        fn heap_size_bytes(&self) -> u64 {
            match self {
                Self::SetStoreInfo(set_store_info) => set_store_info.heap_size_bytes(),
                Self::ArrowMsg(arrow_msg) => arrow_msg.heap_size_bytes(),
                Self::BlueprintActivationCommand(blueprint_activation_command) => {
                    blueprint_activation_command.heap_size_bytes()
                }
            }
        }
    }

    impl SizeBytes for crate::log_msg::v0::SetStoreInfo {
        #[inline]
        fn heap_size_bytes(&self) -> u64 {
            let Self { row_id, info } = self;

            row_id.heap_size_bytes() + info.heap_size_bytes()
        }
    }

    impl SizeBytes for crate::common::v0::Tuid {
        #[inline]
        fn heap_size_bytes(&self) -> u64 {
            let Self { inc, time_ns } = self;

            inc.heap_size_bytes() + time_ns.heap_size_bytes()
        }
    }

    impl SizeBytes for crate::log_msg::v0::StoreInfo {
        #[inline]
        fn heap_size_bytes(&self) -> u64 {
            let Self {
                application_id,
                store_id,
                is_official_example,
                started,
                store_source,
                store_version,
            } = self;

            application_id.heap_size_bytes()
                + store_id.heap_size_bytes()
                + is_official_example.heap_size_bytes()
                + started.heap_size_bytes()
                + store_source.heap_size_bytes()
                + store_version.heap_size_bytes()
        }
    }

    impl SizeBytes for crate::common::v0::ApplicationId {
        #[inline]
        fn heap_size_bytes(&self) -> u64 {
            let Self { id } = self;

            id.heap_size_bytes()
        }
    }

    impl SizeBytes for crate::common::v0::StoreId {
        #[inline]
        fn heap_size_bytes(&self) -> u64 {
            let Self { kind, id } = self;

            kind.heap_size_bytes() + id.heap_size_bytes()
        }
    }

    impl SizeBytes for crate::common::v0::Time {
        #[inline]
        fn heap_size_bytes(&self) -> u64 {
            let Self { nanos_since_epoch } = self;

            nanos_since_epoch.heap_size_bytes()
        }
    }

    impl SizeBytes for crate::log_msg::v0::StoreSource {
        #[inline]
        fn heap_size_bytes(&self) -> u64 {
            let Self { kind, extra } = self;

            kind.heap_size_bytes() + extra.heap_size_bytes()
        }
    }

    impl SizeBytes for crate::log_msg::v0::StoreSourceExtra {
        #[inline]
        fn heap_size_bytes(&self) -> u64 {
            let Self { payload } = self;

            payload.heap_size_bytes()
        }
    }

    impl SizeBytes for crate::log_msg::v0::StoreVersion {
        #[inline]
        fn heap_size_bytes(&self) -> u64 {
            let Self { crate_version_bits } = self;

            crate_version_bits.heap_size_bytes()
        }
    }

    impl SizeBytes for crate::log_msg::v0::ArrowMsg {
        #[inline]
        fn heap_size_bytes(&self) -> u64 {
            let Self {
                store_id,
                compression,
                uncompressed_size,
                encoding,
                payload,
            } = self;

            store_id.heap_size_bytes()
                + compression.heap_size_bytes()
                + uncompressed_size.heap_size_bytes()
                + encoding.heap_size_bytes()
                + payload.heap_size_bytes()
        }
    }

    impl SizeBytes for crate::log_msg::v0::BlueprintActivationCommand {
        #[inline]
        fn heap_size_bytes(&self) -> u64 {
            let Self {
                blueprint_id,
                make_active,
                make_default,
            } = self;

            blueprint_id.heap_size_bytes()
                + make_active.heap_size_bytes()
                + make_default.heap_size_bytes()
        }
    }
}
