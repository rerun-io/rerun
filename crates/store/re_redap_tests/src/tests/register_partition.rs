#![expect(clippy::unwrap_used)]

use arrow::array::RecordBatch;
use futures::TryStreamExt as _;
use itertools::Itertools as _;
use url::Url;

use re_protos::{
    cloud::v1alpha1::{
        CreateDatasetEntryRequest, DataSource, DataSourceKind, ScanDatasetManifestRequest,
        ScanDatasetManifestResponse, ScanPartitionTableRequest, ScanPartitionTableResponse,
        rerun_cloud_service_server::RerunCloudService,
    },
    headers::RerunHeadersInjectorExt as _,
};

use super::common::{DataSourcesDefinition, LayerDefinition, RerunCloudServiceExt as _};
use crate::{RecordBatchExt as _, create_simple_recording_in};

pub async fn register_and_scan_simple_dataset(service: impl RerunCloudService) {
    let data_sources_def = DataSourcesDefinition::new([
        LayerDefinition::simple("my_partition_id1", &["my/entity", "my/other/entity"]),
        LayerDefinition::simple("my_partition_id2", &["my/entity"]),
        LayerDefinition::simple(
            "my_partition_id3",
            &["my/entity", "another/one", "yet/another/one"],
        ),
    ]);

    let dataset_name = "my_dataset1";
    service.create_dataset_entry_with_name(dataset_name).await;
    service
        .register_with_dataset_name(dataset_name, data_sources_def.to_data_sources())
        .await;

    scan_partition_table_and_snapshot(&service, dataset_name, "simple").await;
    scan_dataset_manifest_and_snapshot(&service, dataset_name, "simple").await;
}

pub async fn register_and_scan_simple_dataset_with_layers(service: impl RerunCloudService) {
    let data_sources_def = DataSourcesDefinition::new([
        LayerDefinition::simple(
            "partition1",
            &["my/entity", "another/one", "yet/another/one"],
        ),
        LayerDefinition::simple("partition1", &["extra/entity"]).layer_name("extra"),
        LayerDefinition::simple("partition2", &["another/one", "yet/another/one"])
            .layer_name("base"),
        LayerDefinition::simple("partition2", &["extra/entity"]).layer_name("extra"),
        LayerDefinition::simple("partition3", &["i/am/alone"]),
    ]);

    let dataset_name = "dataset_with_layers";
    service.create_dataset_entry_with_name(dataset_name).await;
    service
        .register_with_dataset_name(dataset_name, data_sources_def.to_data_sources())
        .await;

    scan_partition_table_and_snapshot(&service, dataset_name, "simple_with_layers").await;
    scan_dataset_manifest_and_snapshot(&service, dataset_name, "simple_with_layers").await;
}

pub async fn register_with_prefix(fe: impl RerunCloudService) {
    let root_dir = tempfile::tempdir().expect("creating temp dir");

    // Note: for this test, we don't use teh `DataSourceDefinition` abstraction here because we need
    // tight control of where the RRDs are stored.
    let tuid_prefix1 = 1;
    create_simple_recording_in(
        tuid_prefix1,
        "my_partition_id1",
        &["my/entity", "my/other/entity"],
        root_dir.path(),
    )
    .expect("creating recording");

    let tuid_prefix2 = 2;
    create_simple_recording_in(
        tuid_prefix2,
        "my_partition_id2",
        &["my/entity"],
        root_dir.path(),
    )
    .expect("creating recording");

    let tuid_prefix3 = 3;
    create_simple_recording_in(
        tuid_prefix3,
        "my_partition_id3",
        &["my/entity", "another/one", "yet/another/one"],
        root_dir.path(),
    )
    .expect("creating recording");

    let dataset_name = "my_dataset1";
    fe.create_dataset_entry(tonic::Request::new(CreateDatasetEntryRequest {
        name: Some(dataset_name.to_owned()),
        id: None,
    }))
    .await
    .unwrap();

    let root_url =
        Url::parse(&format!("file://{}/", root_dir.path().display())).expect("creating root url");

    fe.register_with_dataset_name(
        dataset_name,
        vec![
            DataSource {
                storage_url: Some(root_url.to_string()),
                prefix: true,
                layer: None,
                typ: DataSourceKind::Rrd as i32,
            }, //
        ],
    )
    .await;

    scan_partition_table_and_snapshot(&fe, dataset_name, "register_prefix_partitions").await;
    scan_dataset_manifest_and_snapshot(&fe, dataset_name, "register_prefix_manifest").await;
}

// Scanning an empty dataset should return an empty dataframe with the expected schema -- not a
// NOT_FOUND error.
pub async fn register_and_scan_empty_dataset(service: impl RerunCloudService) {
    let dataset_name = "empty_dataset";
    service.create_dataset_entry_with_name(dataset_name).await;

    scan_partition_table_and_snapshot(&service, dataset_name, "empty").await;
    scan_dataset_manifest_and_snapshot(&service, dataset_name, "empty").await;
}

// ---

async fn scan_partition_table_and_snapshot(
    service: &impl RerunCloudService,
    dataset_name: &str,
    snapshot_name: &str,
) {
    let resps: Vec<_> = service
        .scan_partition_table(
            tonic::Request::new(ScanPartitionTableRequest {
                columns: vec![], // all of them
            })
            .with_entry_name(dataset_name)
            .unwrap(),
        )
        .await
        .unwrap()
        .into_inner()
        .try_collect()
        .await
        .unwrap();

    let batches: Vec<RecordBatch> = resps
        .into_iter()
        .map(|resp| resp.data.unwrap().try_into().unwrap())
        .collect_vec();

    let batch = arrow::compute::concat_batches(
        batches
            .first()
            .expect("there should be at least one batch")
            .schema_ref(),
        &batches,
    )
    .unwrap()
    .auto_sort_rows()
    .unwrap();

    let columns = ScanPartitionTableResponse::fields();
    let columns_names = columns
        .iter()
        .map(|field| field.name().as_str())
        .filter(|name| {
            // these are implementation-dependent
            name != &ScanPartitionTableResponse::FIELD_STORAGE_URLS
                && name != &ScanPartitionTableResponse::FIELD_SIZE_BYTES
                // these are unstable
                && name != &ScanPartitionTableResponse::FIELD_LAST_UPDATED_AT
        })
        .collect_vec();
    let filtered_batch = batch.filtered_columns(&columns_names);

    insta::assert_snapshot!(
        format!("{snapshot_name}_partitions_schema"),
        batch.format_schema_snapshot()
    );
    insta::assert_snapshot!(
        format!("{snapshot_name}_partitions_data"),
        filtered_batch.format_snapshot(false)
    );
}

async fn scan_dataset_manifest_and_snapshot(
    service: &impl RerunCloudService,
    dataset_name: &str,
    snapshot_name: &str,
) {
    let resps: Vec<_> = service
        .scan_dataset_manifest(
            tonic::Request::new(ScanDatasetManifestRequest {
                columns: vec![], // all of them
            })
            .with_entry_name(dataset_name)
            .unwrap(),
        )
        .await
        .unwrap()
        .into_inner()
        .try_collect()
        .await
        .unwrap();

    let batches: Vec<RecordBatch> = resps
        .into_iter()
        .map(|resp| resp.data.unwrap().try_into().unwrap())
        .collect_vec();

    let batch = arrow::compute::concat_batches(
        batches
            .first()
            .expect("there should be at least one batch")
            .schema_ref(),
        &batches,
    )
    .unwrap();

    let columns = ScanDatasetManifestResponse::fields();
    let columns_names = columns
        .iter()
        .map(|field| field.name().as_str())
        .filter(|name| {
            // these are implementation-dependent
            name != &ScanDatasetManifestResponse::FIELD_STORAGE_URL
                && name != &ScanDatasetManifestResponse::FIELD_SIZE_BYTES
                // these are unstable
                && name != &ScanDatasetManifestResponse::FIELD_LAST_UPDATED_AT
                && name != &ScanDatasetManifestResponse::FIELD_REGISTRATION_TIME
        })
        .collect_vec();
    let filtered_batch = batch
        .filtered_columns(&columns_names)
        .auto_sort_rows()
        .unwrap();

    insta::assert_snapshot!(
        format!("{snapshot_name}_manifest_schema"),
        batch.format_schema_snapshot()
    );
    insta::assert_snapshot!(
        format!("{snapshot_name}_manifest_data"),
        filtered_batch.format_snapshot(false)
    );
}
