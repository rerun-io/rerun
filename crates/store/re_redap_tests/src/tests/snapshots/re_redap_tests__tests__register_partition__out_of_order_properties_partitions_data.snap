---
source: crates/store/re_redap_tests/src/tests/register_partition.rs
expression: filtered_batch.format_snapshot(false)
---
┌────────────────────┬───────────────────┬──────────────────┬────────────────────────────────────────────────────────┐
│ rerun_partition_id ┆ rerun_layer_names ┆ rerun_num_chunks ┆ property:text_log:TextLog:text                         │
╞════════════════════╪═══════════════════╪══════════════════╪════════════════════════════════════════════════════════╡
│ my_partition_id1   ┆ [prop2, prop1]    ┆ 2                ┆ [I was logged first, registered last, so I should win] │
└────────────────────┴───────────────────┴──────────────────┴────────────────────────────────────────────────────────┘
